### ISSUE TITLE
[BUG] Downloading bundle: .. checksum mismatch as part of `crc setup`

### ISSUE THREAD

**Issue by praveenkumar** [crc-org/crc#3543]
Looks like there is issue with content gateway and due to that redirection is not happening as expected and user can get following error
```
$ crc setup
INFO Using bundle path /Users/yevhen/.crc/cache/crc_vfkit_4.12.5_amd64.crcbundle
INFO Checking if running as non-root
INFO Checking if crc-admin-helper executable is cached
INFO Checking if running on a supported CPU architecture
INFO Checking minimum RAM requirements
INFO Checking if crc executable symlink exists
INFO Checking if running emulated on a M1 CPU
INFO Checking if vfkit is installed
INFO Checking if CRC bundle is extracted in '$HOME/.crc'
INFO Checking if /Users/yevhen/.crc/cache/crc_vfkit_4.12.5_amd64.crcbundle exists
INFO Getting bundle for the CRC executable
INFO Downloading bundle: /Users/yevhen/.crc/cache/crc_vfkit_4.12.5_amd64.crcbundle...
checksum mismatch
```

Edit: We are still working internally to fix those link meanwhile as workaround same bundles are uploaded to gcp bucket and following should work.

- Windows
```
crc setup -b https://storage.googleapis.com/crc-bundle-github-ci/4.12.5/crc_hyperv_4.12.5_amd64.crcbundle
```

- Linux
```
crc setup -b https://storage.googleapis.com/crc-bundle-github-ci/4.12.5/crc_libvirt_4.12.5_amd64.crcbundle
```

- Mac-Intel
```
crc setup -b https://storage.googleapis.com/crc-bundle-github-ci/4.12.5/crc_vfkit_4.12.5_amd64.crcbundle
```

- Mac-M1
```
crc setup -b https://storage.googleapis.com/crc-bundle-github-ci/4.12.5/crc_vfkit_4.12.5_arm64.crcbundle
```
We will update the issue once we have more info around content gateway fix.

**Comment by pbolduc** [comment]
I am also getting this same error,

```
INFO Using bundle path C:\Users\Phil\.crc\cache\crc_hyperv_4.12.5_amd64.crcbundle
INFO Checking if current user is in crc-users and Hyper-V admins group
INFO Checking if CRC bundle is extracted in '$HOME/.crc'
INFO Checking if C:\Users\Phil\.crc\cache\crc_hyperv_4.12.5_amd64.crcbundle exists
INFO Getting bundle for the CRC executable
INFO Downloading bundle: C:\Users\Phil\.crc\cache\crc_hyperv_4.12.5_amd64.crcbundle...
checksum mismatch
```

**Comment by AlphaMalik** [comment]
Same here as well ... came here to report the exact same issue
thought, I was the only one seeing this issue but seems like others are encountering the exact same issue as well.

**Comment by pbolduc** [comment]
Is there a direct link to download the file so we can copy manually into our cache directory until the issue is resolved?

**Comment by pbolduc** [comment]
I tried the URLs [here](https://github.com/crc-org/crc/blob/50b08d5b11d18543f8302c2cda94cb03b70093f0/pkg/crc/machine/bundle/constants.go#LL8C28-L8C28) but I am getting an error,

```
Sorry, it looks like that link has expired.
Download links are dynamic and available for 240 minutes before they expire.

If you’ve followed a bookmark, you will need to browse to the content you want to download and try again.

If you believe you have reached this page in error, contact [Red Hat Customer Service.](https://access.redhat.com/support/contact/customerService)
```

**Comment by TechnoRoss** [comment]
any word or ETA timeline or link?

**Comment by samsenrohm** [comment]
Yes, I am getting the same error as well...

Glad to know that it's not just me... 

And that someone is working one it!

Much appreciated...

D:\MAS8>crc setup
INFO Using bundle path C:\Users\samsen\.crc\cache\crc_hyperv_4.12.5_amd64.crcbundle
INFO Checking if current user is in crc-users and Hyper-V admins group
INFO Checking if CRC bundle is extracted in '$HOME/.crc'
INFO Checking if C:\Users\samsen\.crc\cache\crc_hyperv_4.12.5_amd64.crcbundle exists
INFO Getting bundle for the CRC executable
INFO Downloading bundle: C:\Users\samsen\.crc\cache\crc_hyperv_4.12.5_amd64.crcbundle...
checksum mismatch

D:\MAS8>

And here's the content of crc.log:

time="2023-03-14T15:27:10-06:00" level=debug msg="CRC version: 2.15.0+72256c3c\n"
time="2023-03-14T15:27:10-06:00" level=debug msg="OpenShift version: 4.12.5\n"
time="2023-03-14T15:27:10-06:00" level=debug msg="Podman version: 4.3.1\n"
time="2023-03-14T15:27:10-06:00" level=debug msg="Running 'crc setup'"
time="2023-03-14T15:27:10-06:00" level=info msg="Using bundle path C:\\Users\\samsen\\.crc\\cache\\crc_hyperv_4.12.5_amd64.crcbundle"
time="2023-03-14T15:27:10-06:00" level=info msg="Checking if current user is in crc-users and Hyper-V admins group"
time="2023-03-14T15:27:10-06:00" level=debug msg="Running 'Get-LocalGroupMember -Group 'crc-users' -Member 'samsen''"
time="2023-03-14T15:27:10-06:00" level=debug msg="Running 'Get-LocalGroupMember -SID 'S-1-5-32-578' -Member 'samsen''"
time="2023-03-14T15:27:11-06:00" level=info msg="Checking if CRC bundle is extracted in '$HOME/.crc'"
time="2023-03-14T15:27:11-06:00" level=info msg="Checking if C:\\Users\\samsen\\.crc\\cache\\crc_hyperv_4.12.5_amd64.crcbundle exists"
time="2023-03-14T15:27:11-06:00" level=debug msg="error getting bundle info for crc_hyperv_4.12.5_amd64.crcbundle: could not find cached bundle info in C:\\Users\\samsen\\.crc\\cache\\crc_hyperv_4.12.5_amd64: CreateFile C:\\Users\\samsen\\.crc\\cache\\crc_hyperv_4.12.5_amd64: The system cannot find the file specified."
time="2023-03-14T15:27:11-06:00" level=debug msg="could not find cached bundle info in C:\\Users\\samsen\\.crc\\cache\\crc_hyperv_4.12.5_amd64: CreateFile C:\\Users\\samsen\\.crc\\cache\\crc_hyperv_4.12.5_amd64: The system cannot find the file specified."
time="2023-03-14T15:27:11-06:00" level=info msg="Getting bundle for the CRC executable"
time="2023-03-14T15:27:11-06:00" level=debug msg="Ensuring directory C:\\Users\\samsen\\.crc\\cache exists"
time="2023-03-14T15:27:11-06:00" level=debug msg="Got bundle path: C:\\Users\\samsen\\.crc\\cache\\crc_hyperv_4.12.5_amd64.crcbundle"
time="2023-03-14T15:27:11-06:00" level=info msg="Downloading bundle: C:\\Users\\samsen\\.crc\\cache\\crc_hyperv_4.12.5_amd64.crcbundle..."
time="2023-03-14T15:27:11-06:00" level=debug msg="Downloading https://mirror.openshift.com/pub/openshift-v4/clients/crc/bundles/openshift/4.12.5/crc_hyperv_4.12.5_amd64.crcbundle to C:\\Users\\samsen\\.crc\\cache\\crc_hyperv_4.12.5_amd64.crcbundle"

And when I try to access the above ... mirror ...  link directly...

Sorry, it looks like that link has expired.
Download links are dynamic and available for 240 minutes before they expire.

If you’ve followed a bookmark, you will need to browse to the content you want to download and try again.

If you believe you have reached this page in error, contact [Red Hat Customer Service.](https://access.redhat.com/support/contact/customerService)

**Comment by gbraad** [comment]
> any word or ETA timeline or link?

Sorry for the inconvenience. We have informed the people who are responsible for the content (the CDN), but so far we have not seen a solution. We are considering an alternative. Please keep an eye on this issue.

**Comment by cfergeau** [comment]
In terms of workarounds, if you used crc previously, you can downgrade to the version you were using before, or you can use the latest `crc` release and point it at an older 4.12 bundle in `~/.crc/cache` if you do have one:
```
crc setup --bundle ~/.crc/cache/crc_libvirt_4.12.1_amd64
crc start --bundle ~/.crc/cache/crc_libvirt_4.12.1_amd64
```

If you need/want the latest  bundle (4.12.5), or if you have no cached bundles, we don't have a workaround or a fix yet.

**Comment by praveenkumar** [comment]
We are still working internally to fix those links. Meanwhile as a workaround, the bundles were uploaded to GCP: 

- **Windows**
crc setup -b https://storage.googleapis.com/crc-bundle-github-ci/4.12.5/crc_hyperv_4.12.5_amd64.crcbundle

- **Linux**
crc setup -b https://storage.googleapis.com/crc-bundle-github-ci/4.12.5/crc_libvirt_4.12.5_amd64.crcbundle

- **Mac-Intel**
crc setup -b https://storage.googleapis.com/crc-bundle-github-ci/4.12.5/crc_vfkit_4.12.5_amd64.crcbundle

- **Mac-M1**
crc setup -b https://storage.googleapis.com/crc-bundle-github-ci/4.12.5/crc_vfkit_4.12.5_arm64.crcbundle

sha256 hashes to verify the downloads:
```
f89fc8fa21c762e8e483575950bef5b2aa62e575df6ed0fb76051258ef6352e5  crc_vfkit_4.12.5_arm64.crcbundle
f49945f40c9f219845d2086b66fcc29cbca3b8490bf6319d2a40f4e3e7bd1ad8  crc_hyperv_4.12.5_amd64.crcbundle
12797290dda62930f65e74fd12b8d634ef4d265f50ec708563f37b91ad68edc6  crc_libvirt_4.12.5_amd64.crcbundle
7a691040952b138b200d8bd7dd57ba352ba99f0c4006c976b3660bf7e2d5ccc0  crc_vfkit_4.12.5_amd64.crcbundle
```
We will update the issue once we have more info around content gateway fix.

**Comment by lubomben** [comment]
Thank for temporary fix but unfortunately I get this error on Mac M1:

`INFO Downloading bundle: https://storage.googleapis.com/crc-bundle-github-ci/4.12.5/crc_vfkit_4.12.5_arm64.crcbundle...
0 B / 3.07 GiB [_________________________________________________________________________________________________________________________________________________________________________________________________________] 0.00% ? p/s
INFO Uncompressing /Users/luca/.crc/cache/crc_vfkit_4.12.5_arm64.crcbundle
crc.img:  31.00 GiB / 31.00 GiB [--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------] 100.00%
Temporary error: rename /Users/luca/.crc/cache/tmp-extract/crc_vfkit_4.12.5_arm64 /Users/luca/.crc/cache/crc_vfkit_4.12.5_arm64: no such file or directory (x12)`

**Comment by PingDavidR** [comment]
on Mac Intel, `crc setup` works fine with the flags provided - I was able to download & extract everything.  However, `crc start` yields

> INFO Checking if running as non-root
> INFO Checking if crc-admin-helper executable is cached
> INFO Checking if running on a supported CPU architecture
> INFO Checking minimum RAM requirements
> INFO Checking if crc executable symlink exists
> INFO Checking if running emulated on a M1 CPU
> INFO Checking if vfkit is installed
> INFO Checking if old launchd config for tray and/or daemon exists
> INFO Checking if crc daemon plist file is present and loaded
> INFO Loading bundle: crc_vfkit_4.12.5_amd64...
> Error loading machine: Error loading bundle metadata

**Comment by anjannath** [comment]
@lubomben bundles from https://github.com/crc-org/crc/issues/3543#issuecomment-1469736722 should work now, we had a wrong upload of the mac-M1 bundle, but its **fixed** now

**Comment by anjannath** [comment]
@PingDavidR i was unable to reproduce your issue, the amd64 bundle link is correct and i was able to start CRC with the bundle from https://github.com/crc-org/crc/issues/3543#issuecomment-1469736722

**Comment by lubomben** [comment]
@anjannath I confirm now it works but unfortunately now I have the following issue:

`➜  ~ crc start
Is 'crc daemon' running? Cannot reach daemon API: Get "http://unix/api/version": dial unix /Users/luca/.crc/crc-http.sock: connect: no such file or directory`

**Comment by PingDavidR** [comment]
> > crc setup -b https://storage.googleapis.com/crc-bundle-github-ci/4.12.5/crc_vfkit_4.12.5_amd64.crcbundle
> 
> INFO Using bundle path https://storage.googleapis.com/crc-bundle-github-ci/4.12.5/crc_vfkit_4.12.5_amd64.crcbundle
> INFO Checking if running as non-root
> INFO Checking if crc-admin-helper executable is cached
> INFO Checking if running on a supported CPU architecture
> INFO Checking minimum RAM requirements
> INFO Checking if crc executable symlink exists
> INFO Checking if running emulated on a M1 CPU
> INFO Checking if vfkit is installed
> INFO Checking if CRC bundle is extracted in '$HOME/.crc'
> INFO Checking if https://storage.googleapis.com/crc-bundle-github-ci/4.12.5/crc_vfkit_4.12.5_amd64.crcbundle exists
> INFO Getting bundle for the CRC executable
> INFO Downloading bundle: https://storage.googleapis.com/crc-bundle-github-ci/4.12.5/crc_vfkit_4.12.5_amd64.crcbundle...
> 3.07 GiB / 3.07 GiB [------------------------------------------------------------------------------------------------------------------------------------------->] 99.85% 9.59 MiB p/s
> INFO Uncompressing /Users/davidross/.crc/cache/crc_vfkit_4.12.5_amd64.crcbundle
> crc.img:  31.00 GiB / 31.00 GiB [--------------------------------------------------------------------------------------------------------------------------------------------] 100.00%
> INFO Checking if old launchd config for tray and/or daemon exists
> INFO Checking if crc daemon plist file is present and loaded
> Your system is correctly setup for using CRC. Use 'crc start' to start the instance
> 
> ╰$ crc start
> INFO Checking if running as non-root
> INFO Checking if crc-admin-helper executable is cached
> INFO Checking if running on a supported CPU architecture
> INFO Checking minimum RAM requirements
> INFO Checking if crc executable symlink exists
> INFO Checking if running emulated on a M1 CPU
> INFO Checking if vfkit is installed
> INFO Checking if old launchd config for tray and/or daemon exists
> INFO Checking if crc daemon plist file is present and loaded
> INFO Loading bundle: crc_vfkit_4.12.5_amd64...
> Error loading machine: Error loading bundle metadata


Open to suggestions

**Comment by PingDavidR** [comment]
```
CRC version: 2.15.0+72256c3c
OpenShift version: 4.12.5
Podman version: 4.3.1
```

**Comment by lubomben** [comment]
> @anjannath I confirm now it works but unfortunately now I have the following issue:
> 
> `➜ ~ crc start Is 'crc daemon' running? Cannot reach daemon API: Get "http://unix/api/version": dial unix /Users/luca/.crc/crc-http.sock: connect: no such file or directory`

Sorry, `crc cleanup && crc setup` resolved the issue, now crc is started

**Comment by PingDavidR** [comment]
Ah, borrowing from @lubomben, I did a `crc cleanup / crc setup & crc start `- my pull secret needed to be re-entered, and I was prompted this time. It is working now.  Thanks!

**Comment by andysmagana** [comment]
Still getting the checksum mismatch error from the OpenShift Local Install from Red Hat
INFO Downloading bundle: /home/asmagana/.crc/cache/crc_libvirt_4.12.5_amd64.crcbundle... 
checksum mismatch
[asmagana@rh87 ~]$

**Comment by cfergeau** [comment]
> Still getting the checksum mismatch error from the OpenShift Local Install from Red Hat INFO Downloading bundle: /home/asmagana/.crc/cache/crc_libvirt_4.12.5_amd64.crcbundle... checksum mismatch [asmagana@rh87 ~]$

Yes, we're still waiting on the team managing this download location to look into our issue, so you still need this workaround https://github.com/crc-org/crc/issues/3543#issuecomment-1469736722

**Comment by pbolduc** [comment]
> Still getting the checksum mismatch error from the OpenShift Local Install from Red Hat INFO Downloading bundle: /home/asmagana/.crc/cache/crc_libvirt_4.12.5_amd64.crcbundle... checksum mismatch [asmagana@rh87 ~]$

You will need to use the work around from above to specify an alternative download location when running setup.

**Comment by pbolduc** [comment]
> We are still working internally to fix those links. Meanwhile as a workaround, the bundles were uploaded to GCP:
> 
> * **Windows**
>   crc setup -b https://storage.googleapis.com/crc-bundle-github-ci/4.12.5/crc_hyperv_4.12.5_amd64.crcbundle
> * **Linux**
>   crc setup -b https://storage.googleapis.com/crc-bundle-github-ci/4.12.5/crc_libvirt_4.12.5_amd64.crcbundle
> * **Mac-Intel**
>   crc setup -b https://storage.googleapis.com/crc-bundle-github-ci/4.12.5/crc_vfkit_4.12.5_amd64.crcbundle
> * **Mac-M1**
>   crc setup -b https://storage.googleapis.com/crc-bundle-github-ci/4.12.5/crc_vfkit_4.12.5_arm64.crcbundle
> 
> We will update the issue once we have more info around content gateway fix.

This worked for me on Windows and Hyper-V.

**Comment by andysmagana** [comment]
I am getting this error after crc start -p pull-secret 

time="2023-03-15T14:40:42-05:00" level=debug msg="SSH command succeeded"
time="2023-03-15T14:40:42-05:00" level=debug msg="error: Temporary error: pull secret not updated to disk - sleeping 2s"

**Comment by pbolduc** [comment]
> I am getting this error after crc start -p pull-secret
> 
> time="2023-03-15T14:40:42-05:00" level=debug msg="SSH command succeeded" time="2023-03-15T14:40:42-05:00" level=debug msg="error: Temporary error: pull secret not updated to disk - sleeping 2s"

Did your bundle get downloaded correctly. If so, the issue you are experiencing sounds like a different issue. I would validate the bundle was downloaded and extracted successfully. If it was, I would recommend you create a separate issue with more details.

**Comment by cfergeau** [comment]
The initial issue should be resolved now, `crc setup` should be able to download the bundle automatically, and not error out with `checksum mismatch`

**Comment by praveenkumar** [comment]
Closing this issue since it is fixed on CGW side.

**Comment by gbraad** [comment]
We are looking into a workaround method to prevent this in the future: #3545

**Comment by it-pappa** [comment]
I have downloaded the latest Openshift local, but when running "crc setup" I get: 
`unable to get verified hash for default bundle: Get "https://mirror.openshift.com/pub/openshift-v4/clients/crc/bundles/openshift/4.12.13/sha256sum.txt.sig": context deadline exceeded (Client.Timeout exceeded while awaiting headers)`


When trying the workaround above: I get 403 forbidden.

**Comment by anjannath** [comment]
@it-pappa this issue is fixed now, the error you are getting is not directly related, it hit a timeout while trying to download the bundles sha256 hashes, most probably due to slow internet. 

most of the time it'll go away if you run `crc setup` again

**Comment by it-pappa** [comment]
@anjannath hi, I have 250/250 fiber connection with unlimited data. Network is not a problem. I fixed it by manually download the bundle and run "crc setup -b /path/to/downloaded-file"

I saw others also had that issue on reddit.

**Comment by anjannath** [comment]
@it-pappa yeah, it could be the shorter timeout, see: https://github.com/crc-org/crc/issues/3686#issuecomment-1568363710
